### PR TITLE
Add models to store CKAN query configuration for a resource

### PR DESCRIPTION
**Pull request by NightJar**
In order to present data extracted from a CKAN resource, we'll need
to store the configuration to form a URI to request against the API.
This means we need data models, and a way to edit them in the CMS.

### CODE REVIEW

File: src/Model/ResourceField.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            ->setAttribute('placeholder', $this->Name);
+
+        $summary = $fields->dataFieldByName('ShowInSummaryView');
+        $detail = $fields->dataFieldByName('ShowInDetailView');
+        $duplicates = $fields->dataFieldByName('RemoveDuplicates');
+
+        $fields->removeByName(['ShowInSummaryView', 'ShowInDetailView', 'RemoveDuplicates',]);
+        $fields->insertBefore(FieldGroup::create('Visibility', [$summary, $detail, $duplicates]), 'Order');
+
+        $fields->removeByName('ResourceID');
```
Comment (robbieaverill) [inline src/Model/ResourceField.php]: Minor comment, but it's always a good idea to use `beforeUpdateCMSFields` in this case to ensure extensions can modify these fields

File: src/Model/ResourceField.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        'Order',
+        'ReadableName',
+        'Name',
+        'Type',
+        'ShowInSummaryView',
+        'ShowInDetailView',
+    ];
+
+    /**
+     * Always display the 'ReadableName' unless it's unset, then display the name that is returned by CKAN
```
Comment (robbieaverill) [inline src/Model/ResourceField.php]: I'd imagined this being an `onBeforeWrite()` call to build and set the readable name. Perhaps we could adjust this in a future PR

File: src/Model/ResourceFilter.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            $typeTitle,
+            $this->config()->get('filter_types')
+        ));
+        $fields->replaceField('TypeOptions', HiddenField::create('TypeOptions'));
+        $fields->push(ListboxField::create(
+            'FilterFields',
+            'Columns to search',
+            $this->FilterFor()->Fields()->map('ID', 'ReadableName')
+        ));
+        $fields->removeByName('FilterForID');
```
Comment (robbieaverill) [inline src/Model/ResourceFilter.php]: As with the previous model, it'd be best to put this logic into `beforeUpdateCMSFields()`

File: src/Model/ResourceFilter.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        return $fields;
+    }
+
+    public function forTemplate()
+    {
+        $options = json_decode($this->TypeOptions, true);
+        $field = Injector::inst()->createWithArgs($this->Type, $options);
+        if ($field instanceof FormField) {
```
Comment (robbieaverill) [inline src/Model/ResourceFilter.php]: This logic looks backwards to me. Should it not be the opposite? If the field is NOT a FormField, throw that exception?

File: src/Page/CKANRegistryPage.php [hunk 5]
```diff
@@ -29,8 +36,18 @@ class CKANRegistryPage extends Page
 
     public function getCMSFields()
     {
-        $this->beforeUpdateCMSFields(function (FieldList $fields) {
-            $fields->addFieldToTab('Root.Data', ResourceLocatorField::create('Thing'));
+        $resource = $this->DataResource();
```
Comment (robbieaverill) [inline src/Page/CKANRegistryPage.php]: `$this` is bound to closures in PHP 5.4 onwards, you don't need to pass it it into the closure like this

**Comment by robbieaverill** [comment]
I've made a PR to address all of these points at https://github.com/silverstripe/silverstripe-ckan-registry/pull/40

File: src/Model/ResourceField.php (hunk 1 above)
Comment (NightJar) [reply src/Model/ResourceField.php]: Reasonable. Thanks for the reminder, old habits die hard :>

**Pull request merged**